### ISSUE TITLE
Urgent Bug: Overselling issue

### ISSUE THREAD

**Issue by mogmachine** [eventespresso/event-espresso-core#109]
Sorry, I wouldn't normally bother devs here but support have told me to do so.

In short, EE4 is overselling all of our events regardless of what limits we set because they are in high demand and multiple people add to the cart at the same time. I have reduced the session to 10 mins, but we oversold 10 tickets by 2 last night which caused a huge issue. 
In 48 hours we have a show with 3000 tickets to go online which will sell out in hours, ...with this issue it has the potential to oversell by thousands.

This is my ticket explaining the issue: https://eventespresso.com/topic/urgent-how-are-capacities-calculated-system-is-overselling-tickets/

..and another ticket which seems to have the same issue and a proposed solution: http://eventespresso.com/topic/timer-to-prevent-overselling/

If we cannot fix this asap I am going to have to take down a live site for which I have been paid to develop and an entire business will fold because there is not time to rebuild using another platform in 48 hours and they are relying on this big sale for revenue.

I am willing to put my hand in my pocket and pay whatever your day rates are to fix this as I have no choice, ..I have recommended and implemented an event ticketing system to my clients and the system cannot keep to the most basic limitations of ticket sales.

Please please please help.

**Comment by tn3rb** [comment]
@mogmachine 

I'm working on a fix right now... 
will post a branch for you to try out the moment it's ready

**Comment by mogmachine** [comment]
@tn3rb  thanks Brent.

**Comment by tn3rb** [comment]
you can take a look at the work I have in progress if you like
https://github.com/eventespresso/event-espresso-core/tree/BUG-8698-ticket-sellouts
as is it should prevent overselling
but requires some further work to release reserved tickets for any failed transactions. 
just haven't pushed those commits yet.
plz DO NOT activate this branch on a production server just yet
no testing has been performed yet
if you have a testing server, then plz feel free to activate this branch there and help out with the testing.

**Comment by mogmachine** [comment]
Thanks B. ..I really appreciate your time and work, ...I will not deploy this to the live site just yet ;) ..and as it is Friday night here I will be out for the next 3-4 hours, ..but then back in front of the computer after that for some potentially beer fuelled testing.
Will check back to see if you have committed the changes by then.

**Comment by tn3rb** [comment]
might want to hold off on your beer fuelled testing, there's a few bugs i need to work out... 
I'll let you know when I push some fixes

**Comment by tn3rb** [comment]
plz feel free to test out that branch

**Comment by mogmachine** [comment]
I have the branch, ..but am shooting a wedding today, ..so will not be able to test until Sunday, ..but I have marked a few hours on Sunday for testing with the client and will report back.

Can you just clarify at which point is should stop the oversell ...when adding to the cart (so not possible to add if tickets are already in others cart)? I assume this, ..just checking so I know what to look out for in the testing.

**Comment by mogmachine** [comment]
First test not good I am afraid. This is what I did.

- made an event with a limit of 5.
- made a ticket with a limit of 2 for this event (often the case that the ticket limit is less than the event limit as there may be multiple ticket types).
- opened 2 new sessions in parallel and added 1 ticket to one cart and 2 to the other.
- went through the reg process in parallel one ofter the other.
- both went through fine and I have oversold my 2 places.
- if I have of opened 50 sessions then I would have oversold by 50.

http://grab.mogmachine.com/JOIm

**Comment by mogmachine** [comment]
It continues.
- I then created a second ticket type with a limit of 3
- on the front end only 2 showed as available (because 3 of the total event capacity of 5 had been sold).

I then in different browsers added those 2 to my cart and checked out simultaneously and again oversold the ticket, ..now overselling the event total too by 2.

http://grab.mogmachine.com/JOII

So all in all, ..didn't work.

I am reverting back to the stable release of EE.

Please advise on if you would like me to test anything else.

**Comment by tn3rb** [comment]
@mogmachine 

sorry to hear you are having troubles. I would like to clarify a few things first (and my apologies in advance if this sounds condescending in any way, but it is standard practice in these types of situations to confirm some of the basics before proceeding).

 1) Please check the version of your Event Espresso installation that you are testing. Is it version 4.8.6.rc.000? You say you have the branch, but we need to confirm this before anything else. Are you using Git or did you download a Zip? If you are using Git, please ensure that you have pulled the latest commits into your local repo. If you are downloading a Zip file, then any time there are changes you have to re-download and install again. I just bumped the version for the branch to 4.8.6.rc.001 so that we can be absolutely certain that you have the latest changes, so please confirm that you are able to install and activate version 4.8.6.rc.001. If the Event Espresso version on the WordPress plugins page does not say 4.8.6.rc.001, then you don't have the fixes in this branch. 

2) When you say that you "opened 2 new sessions in parallel" did you use two tabs in the same browser, or did you use two different browsers (ie: Chrome + Firefox). Opening two or more tabs in the same browser does not actually initiate new sessions, so your tabs will effectively be sharing the same session across those tabs. This won't allow you to oversell the event, but can cause confusion when testing because one tab's actions can overwrite the actions of another tab.

3) Are you also using the Event Espresso Multi-Event Registration (MER) addon ? I am still working on a few fixes for that addon (which should hopefully be completed today), so for the time being, please de-activate that addon.

4) I set up this branch on one of my public testing servers and created a test event that uses the same ticket and datetime settings that you displayed in your screenshots. You can view the event here: 
http://brent.eventespresso.com/events/mog-machine-event/
I have been unable to proceed to registration as soon as one browser selects the only tickets available. Please let me know if you are able to successfully register for more than two of the first ticket (Ticket Limit 2).

To assist you in your testing, there are a couple of filters that you can employ to modify the registration process. The first will display a countdown timer showing how long you have to complete the registration process. It can be activated by adding the following to your theme's functions.php file, or to a custom plugin if you add site functionality that way:

```
// display registration time limit
add_filter( 'FHEE__registration_page_wrapper_template__display_time_limit', '__return_true' );
```

The next filter allows you to change your session expiration time,  which is handy on sites that sell a huge volume of tickets in a short amount of time, because it reduces how long tickets are reserved before people have to complete the registration process. If they take too long, or abandon the registration process, then this allows the reserved tickets to be released and made available again for purchase. The default session expiration is 2 hours, but this filter allows you to set it to whatever you want:

```
function espresso_set_session_lifespan() {
    return 5 * MINUTE_IN_SECONDS;
}
add_filter( 'FHEE__EE_Session__construct___lifespan', 'espresso_set_session_lifespan' );
```

simply change the number to however many minutes you want the session to be.

So the most important thing out of all of the above is to please confirm your Event Espresso plugin version that you are testing. Let me know how that goes.

**Comment by mogmachine** [comment]
@tn3rb  Not in anyway condescending, ..but you will need to forgive me for not being able to reply and further text immediately. I am limited internet for the next couple of days.

**Comment by mogmachine** [comment]
Ok, .finally found some time and a good connection, ..so here goes.

1) Check: http://grab.mogmachine.com/JVdK

2) Have been using incognito windows in Chrome which usually have a new session for each, and also and now primarily, different browsers.

3) Was using it, now disabled.

4) Seen, tested and noted that it works.

Confirmed, .it is working. so I tried to turn on the MER add-on again, ..not working. For now we can deal with not having the MER although it is nice to have the cart and it is something we want in the future so keep me updated how fixing bugs here go.

Also please let me know if this will be committed to the next stable version and when so I don't accidentally lose it in an update.

Thank you so much again, ...just in the nick of time as the client has a sensitive event going on sale tomorrow.

**Comment by tn3rb** [comment]
Ya I will update this ticket when this gets merged into master.

as an FYI, we've also noticed that there are some issues with duplicating events and reserved tickets, so maybe stay away from that for a bit as well.

**Comment by tn3rb** [comment]
I've pushed some fixes to BUG-8698-ticket-sellouts (version 4.8.6.rc.002)
as well as fixes for MER to branch BUG-8698-ticket-sellouts (version 2.0.5.rc.000) at https://github.com/eventespresso/ee4-mer/tree/BUG-8698-ticket-sellouts

**Comment by mogmachine** [comment]
Hi Brent, 

Sorry was out of contact for a week but back in the office now. I am going to update to 4.8.6.rc.002 but am interested to know when this will be committed to a production update as I am uneasy running this on a live site and my client doesn't like the constant reminders that the license has expired.

I am going to leave the MER branch for now until it is committed to production version as 1) your link does not work and 2) I am concerned and there are lots of live events at the moment that I do not want to break.

**Comment by joshfeck** [comment]
It will be merged to Master when testing is complete, and we generally take more than the usual amount of time to test tickets like this one. We also merge the latest changes to Master into this branch so it's for the most part in sync with the next production update.

You can disable the Plugin update engine (which adds the notices about an expired license when it's on a release candidate) by adding the following code to your functions file:

`add_filter( 'FHEE__EE_System_brew_espresso__load_pue', '__return_false' );`

MER is a private repo, but I can ask one of the admins to make it accessible for your account.

**Comment by mogmachine** [comment]
Can you please confirm or deny if this functionality it part of latest 4.8.10 version of EE.

Thanks.

**Comment by joshfeck** [comment]
There is still work left to do on this before it can be merged to Master branch and released. One of the issues that needs to be worked out is making sure an abandoned ticket is added back to what's available. For example, if someone starts a registration for 5 tickets, then gets forwarded to PayPal.com to pay, then has second thoughts and decides to close their browser window and walk away, those 5 tickets need to be added back to the inventory of available tickets.

**Comment by mogmachine** [comment]
I had experienced exactly that issue and had to clear out incomplete transactions by hand on a couple of events.

If you remember when it is committed to let me know, ..else I will keep checking the changelogs and refrain from updating. the current version I am running seems to be working well so far apart from the above.

**Comment by joshfeck** [comment]
It turns out we've been keeping this branch up to date with Event Espresso core master by merging Master into this branch. So if you want to have the latest fixes from EE core, you can pull this branch.

**Comment by salvatorious** [comment]
Is this still an issue with the current version of the plugin? We're about to use it for a large project, but overselling events would be a dealbreaker.

**Comment by garthkoyle** [comment]
It is much more difficult for an event to oversell now that the branch mentioned above (https://github.com/eventespresso/event-espresso-core/tree/BUG-8698-ticket-sellouts) was merged to master on November 29th. It was kind of rare for an event to oversell in the first place (except in very high/active registrations), but that development helps significantly still.

**Comment by salvatorious** [comment]
Great, thanks Garth.

**Comment by mogmachine** [comment]
That said ...I am sorry to say it but @salvatorious I would not use EE for a high traffic site with many many events each with multiple ticket types, ...it gets super buggy on numbers it sells .... check this out from a recent event: http://grab.mogmachine.com/1d2J0D0U322x

50% straight out oversold, and those that did not oversell on sales, did not bother counting the number of registrants as tickets ...so you have 120 tickets, ..it makes 109 sales, ..which uses up 149 tickets. ...this caused us a huge headache as it was to a sold out hip hop show in London at a venue with strict limits.

 ....AND it is not quick under high loads, ...unless you have a dedicated server for the site at a hefty cost (especially the admin).

**Comment by garthkoyle** [comment]
@mogmachine which version and/or branches are you using; you should be using 4.9.23.

Also, can you give expanded views of the tickets so we can see the associated datetimes?

And, how quickly did these registrations come in; over what period of time?

**Comment by nerrad** [comment]
This issue was fixed as of [EE version 4.9.23](https://github.com/eventespresso/event-espresso-core/releases/tag/4.9.23.p).  For users of the Event Espresso Multiple Event Registration add-on, version 2.0.11 of that add-on is required as well for the fix.

Any further issues discovered related to overselling while using those latest versions can be reported in a new ticket.

Thanks for all the help with reporting and testing in here!
